README.md
# Timbits
Widget framework based on Express

## Introduction

Timbits is an attempt to build an easy, and reusable widget framework on top of Express.  These widgets are meant to render independent HTML snippets based on REST based JSON/XML data sources, and brought together on a page via ESI (Edge Side Includes), or our own proprietary CSI (Client Side Includes).

It's primarily meant to serve internal purposes as Postmedia Network Inc, however, it is being open sourced under the MIT License.  Others may find some use for what we are doing, and still others may be able to help turn this into a more generic, and useful solution via contributions.

Constructive criticism is encouraged.

## Installing

Just grab [node.js](http://nodejs.org/#download) and [npm](http://github.com/isaacs/npm) and you're set:

	npm install timbits -g
	
Timbits is simplistic, and finds most of it's power by running on top of some very cool node libraries such as [Express](http://expressjs.com/) and [Hogan](http://twitter.github.com/hogan.js/).  It also uses our [Pantry](https://github.com/Postmedia/pantry) package for efficient utilization of JSON/XML data sources.

## Changes

We completed a significant rewrite of Timbits, these changes occur between version 0.6.7 and 0.7 release of timbits.  For those using previous versions of Timbits here is a quick rundown of the changes.

* major rewrite of Timbits source to JavaScript (formerly CoffeeScript)
* generates JavaScript files by default (CoffeeScript is optional)
* default view engine is now Hogan (formerly CoffeeKup)
* dynamic helpers are no longer supported
* sessions are no longer enabled by default
* uses Winston for logging (formerly coloured-log)

I'll be making a post on my [blog](http://mred9.com) providing more details on these changes, and the reasoning behind them.

Please be aware that the 0.7.x series should be considered experimental, and not for production use.  Once we've stabilized this rewrite it will be released as v0.8.x

### Running previous versions on 0.7

Projects using timbits prior to v0.7 based on CoffeeScript, and Coffeecup require a few changes in order to run. The instructions here provide the details required to get a simple timbit functional.  

There are two specific modifications required, the first is to include a dependency for CoffeeScript, and Coffeecup in your timbits package.json file:

	"dependencies": {
		"coffee-script": "~1.3.3",
		"coffeecup": "~0.3.20",
		..... additional dependencies here
	}

The second change is to the server.js file, you need to do three things:
* require coffee-script
* specify the engine parameter when instantiating the server object
* set the rendering method to handle coffeecup templates. 

The structure of the main server.js file could look like this:

	require('coffee-script');
	
	var timbits = require('timbits');
	var server = timbits.serve({ engine: 'coffee' });

	server.engine( 'coffee', require('coffeecup').__express );

That is all it will take, unless of course you are using dynamic view helpers, for that we need to make a change to the individual timbits themselves.  

Let's suppose you had a helper file helpers/headlines.coffee in your timbits project, in versions prior to 0.7, timbits dynamically loaded all helpers in the 'helpers' path.  These helpers were automatically loaded into the context based on the helper filename, so the example mentioned would load any helper functionality under context.headlines, now you must do this manually.

Include your helpers in each timbit that requires them, at the start of the fetch method. Assuming for example the helper file is headlines.coffee, you would use:

	@fetch req, res, context, options, =>
		# Add helpers manually at beginning of fetch
		context.headlines =  require("../helpers/headlines.coffee");


## Using

The structure of a Timbits application is fairly simple.  To start, you need a two lines in your main server.js file.  

	var timbits = require('timbits');
	var server = timbits.serve();

Timbits, like many other frameworks, prefers convention over configuration.  It will look for specific folders upon startup, and if you manually create your project (vs using the Timbits command line to generate a new project) you should create the following folders.

* /public - images, javascript, stylesheets etc.
* /timbits - this is where we place the individual timbit (widget) files
* /views - views for a particular timbit are placed in a subfolder of the same name

When you start the server, it will automatically load all the timbits found in the /timbits folder.  The name of the timbit is determined by the name of the file, and that name in turn determines the default route (/name/:view?), and the default view (/views/name/default.coffee).  Aside from the location, and name of the timbits, the rest is customizable as shown in the examples.

Timbits can be created in JavaScript (default), or CoffeeScript.  This documentation will assume JavaScript.

The simplest of timbits takes the following form:

	// Load the timbits module
	var timbits = require('../../lib/timbits');

	// create and export the timbit
	var timbit = module.exports = new timbits.Timbit();

That's it!  If you created this as /timbits/test.js, and placed a default view at /views/test/default.hjs, you could "eat" this timbit by going to /test in a browser.  See the "plain" timbit for an example of this.

## How it works

When you start up a Timbits based project, the framework will dynamically load each timbit from the /timbits/ folder.  A timbit is
nothing more than a .js, or .coffee file which creates, and exports a new instance of the Timbit class.

This Timbit class has one very important method named "eat" which is the entry point for every request.  This method is intended to
be overwritten.   When called, the framework will pass in the following three variables

* req - the http request object
* res - the http response object
* context - a hash of data properties

The req, and res parameters come straight from Express.js and are documented [here](http://expressjs.com/guide.html).
The context object originates within Timbits, and will initially contain the following properties:

* name - the name of the timbit being executed
* view - the name of the specified view
* maxAge - the number of seconds the response should be cached for

It's important to note that it is possible to override the view, and/or the maxAge during the execution of the request if needed.

Each timbit by default supports HTTP GET method, it is possible to support one of the alternativce standard HTTP methods as defined in [RFC 2616](http://www.w3.org/Protocols/rfc2616/rfc2616-sec9.html), specifically we support "GET", "POST", "PUT", "HEAD", and "DELETE".  To specified the methods you want to allow (beyond "GET") define the methods as json properties in the timbit.methods property like this

	timbit.methods = { 'get': false, 'post': true }

Values are boolean, the methods are not case sensitive, you only need to specify the methods you wish to override from the default settings. Any request for your timbit with a method not specified will result in an HTTP 405 "Method Not Supported" error.

Each route, querystring, and post parameter is also added to the context as a property.  So, if one included ?page=5 in the request url this value would be made available to you as context.page

Within the eat method, the simplest of implementations will merely call the render method to, well, render the given view using the data found with the context.  The render method takes the same three parameters originally passed into the eat method.  Within the render method Timbits will pass the context to the view rendering engine so that those values can be 
referenced within the view.  So a very simple eat method could look like this

	timbit.eat = function(req, res, context) {
	  this.render(req, res, context);
	};

In fact, that's exactly what it looks like within the Timbit class, so if you don't override it, the timbit will simply pass on the request parameters for rendering.  But that's generally not very useful.  The intent in most cases is to add additional data to the context, whether it be generated on the local server, like a timestamp (see the cherry timbit example), or more likely from an external resource like a RSS, or JSON feed/API.

In most cases you'll need to pull in remote data, so we've incorporated [Pantry](https://github.com/Postmedia/pantry) into Timbits to make fetching (and caching) those resources extra easy.  This is available to you via the built in fetch method.

The fetch method takes five parameters.  The first three are the same as the render method: req, res, context.  It then adds two additional parameters

* options - This is a hash of options passed directly to Pantry's fetch method.  It identifies the source URI of the resource you want to retrieve along with a host of other options as documented within Pantry
* callback - This is an optional parameter which identifies the function to execute once the request has been completed.  It defaults to the render method

The fetch method will automatically add the retrieved resource to context.data by default.  If that isn't acceptable, you can also add a property called "name" to the options parameter to specify an alternate property name to store the data in.  This is shown in the included "chocolate" example as follows:

	// specify the data source
	var src = {
	  name: 'ip',
	  uri: "http://jsonip.com?q=" + context.q + "&rpp=" + context.rpp
	};

	// use the helper method to fetch the data
	// timbit.fetch will call timbit.render once we have the data
	timbit.fetch(req, res, context, src);
	
Be aware that if (using the example above) context.ip already exists, timbits will transform context.ip into an array (if it isn't already), and append the retrieved resource to the array.

Once the request has been completed, the data added to the context, and no callback is specified, the fetch method will then call the render method for you.  If you need to request some additional resources, or perhaps further manipulate the resource once it's been fetched, simply tack on the optional callback parameter, and implement your follow up code.  Timbits will pass on the req, res, and context parameters to the callback.  Just be sure to call the render method once you're done.

	timbit.fetch(req, res, context, src, function(req, res, context){
		// run your custom code here
		// ......
		// ......
		
		// and then call render
		timbit.render(req, res, context);
	});

## Additional Features

### Command Line

Create a new project, generate a timbit (and default view), or run the project with runjs.  Run timbits -h for complete list of commands, and options available.

	timbits n[ew] [project]
	timbits g[enerate] [timbit]
	timbits s[erver] [filename]
	timbits t[est]
	
### Running timbits

There are a number of ways to run timbits, but the easiest approach during development is to utilize the timbits server command.  Doing so will by default run the server.js file, although an alternate filename (server.coffee) can be given.  The timbits server command has the added benefits of loading up any environment variables from an optional .env file, as well as automatically restarting the server if any files change.

### Server Configuration

The server object can be configured during initialization as needed.  The following is a list of possible configuration parameters and their [default value]

* appName - Friendly name of the timbits application ['Timbits']
* engine - Default view engine ['hjs']
* base - A base path for nested servers.  e.g. '/my/nested/server' ['']
* port - The server port to listen on if PORT or C9_PORT env labels are not present. [5678]
* home - The physical path to the root project folder [process.cwd()]
* maxAge - The default cache length in seconds [60]
* discovery - Enables you to serve out your library definition via /timbits/json [true]
* help - Enables the automated help pages via /timbits/help and /[name]/help [true]
* test - Enables the automated test pages via /timbits/test and /[name]/test
* json - Enables the built in json view for timbits via /[name]/json [true]
* jsonp - Enables jsonp requests to json resources

The following example shows how to run a nested application using the [coffeecup](https://github.com/gradus/coffeecup) view engine in place of Hogan and disables the automated test routes.

	var timbits = require('timbits');
	var server = timbits.serve({appName: 'My Nested Timbits', engine: 'coffee', base: '/nested/app', test: false});
	server.engine('coffee', require('coffeecup').__express);
	console.log("Press Ctrl+C to Exit");

### Parameter declaration and validation

For each timbit, you can define a list of parameters which are automatically validated during execution.  This also powers the automated help, test, and discovery functions (see further below).  Parameter attributes you can manipulate are:


History.md
0.7.3 / 2014-12-08
==================
  * Removed use of twitter API v1.0 in templates generated by 'timbits g' and README or examples
  * Added support for all methods on timbit endpoints, default support for GET enabled for backward compatibility
  * Example timbit added to demonstrate http POST method use
  * Testing updated to check and use GET/POST depending on which the timbit supports.
  
0.7.2 / 2014-05-12
==================
  * Dependency update to allow pantry v0.7.x
  * Dependency update to allow request v2.36.x
  * Dependency update to allow express v3.10.x
  * Dependency update to allow hjs v0.0.6
  * Dependency update to allow hogan.js v3.0.x
  * Dependency update to allow run v1.2.x
  * Dependency update to allow mocha v1.20.x
  * Dependency update to allow should v4.0.x
  * Examples - added parameter and example help for dutchie timbit
  * Added dependency for should-http v0.0.2 augments should v4.x
  * Updated node engine support from 0.8.x to up to 0.10.x
  * Updated npm to 1.x 
  * Changed test to require '-c' or '--coffee' to use coffee-script

0.7.1 / 2013-07-16
==================
  * Dependency update to allow pantry v0.5.x
  * Now uses Wordpress API instead of Twitter for examples

0.7.0 / 2013-06-05
==================
  * 0.7.x series should be considered unstable.  stable version will be released as 0.8.x
  * major rewrite of Timbits in JavaScript (formally CoffeeScript)
  * generates JavaScript files by default (CoffeeScript is optional)
  * default view engine is now Hogan (formally CoffeeKup)
  * dynamic helpers are no longer supported
  * sessions are not longer enabled by default
  * now uses Winston for logging (formally coloured-log)
  * updated documentation to reflect changes

0.6.7 / 2012-11-28
==================
  * added trim function when processing environment key/value pairs, fixes issue on Windows when multiple environment variables are concatenated in same string.

0.6.6 / 2012-11-08
==================
  * renamed csi libraries to match best practices (dot instead of dash)

0.6.5 / 2012-10-23
==================
  * client rendering library revised.  no longer uses namespaced includes

0.6.4 / 2012-10-23
==================

  * Add support for responsive client side includes via media attribute
  * Modified production flags to return 404 when feature has been disabled

0.6.3 / 2012-10-22
==================

  * Supports new configuration options which allow you to disable automated discovery, help, tests, and json views.
  * no longer uses our own custom jsonp-filter package

0.6.2 / 2012-10-22
==================

  * Rewrite of timbits command line in js
  * No longer requires coffee-script, mocha or runjs packages to be installed globally
  * for projects created via "timbits new", generated package.json will lock down the timbits dependency to the version used to create the project.  To upgrade to a newer version of timbits, you'll need to adjust the dependency first
  * Added support for numerous mocha options to "timbits test" command

0.6.1 / 2012-10-16
==================

  * Updated read me to reflect global dependencies for command line use of timbits
  * Modified spawn of child process to use the node-which lib for proper PATHEXTS of executables on Windows and Linux

0.6.0 / 2012-10-05
==================

  * Updated dependencies to latest libraries
  * Now requires node v0.8.x
  * timbits are now loaded before server is started (instead of async)
  * Switch to mocha for testing (was using vows)
  * Completely revised automated testing with option for more extensive tests
  * Added test option to timbits command line
  * timbits s[erver] and t[est] will now load environment variables from a .env file if present

0.5.14 / 2012-10-05
===================
  * Removed connect -esi, -assets, -less as requirements (but you can still add them to your project manually if needed)

0.5.13 / 2012-10-05
===================

  * Lock down dependency versions prior to v0.6.x

0.5.12 / 2012-06-28
===================

  * Added alias name support to timbit parameters

0.5.10 / 2012-06-07
===================

  * Added error handling to fetch helper method

0.5.9 / 2012-05-28
==================

  * Better handling of errors when eating or rendering
  * Removed auto-capitalization css from help views

0.5.8 / 2012-05-24
==================

  * Corrected method of returning remove render JSON

0.5.7 / 2012-05-23
==================

  * Added support for onload event to client side esi/csi processing
  * Allow for $(QUERY\_STRING{'name'}) syntax in client side esi/csi processing

0.5.6 / 2012-05-22
==================

  * Added ability to access to pantry directly from within a Timbit

0.5.5 / 2012-05-18
==================

  * Added the ability render timbits client side via JSONP and callback parameter
  * Removed the automatic insert of pantry URIs into context due to security concerns

0.5.4 / 2012-04-25
==================

  * Added the ability to specify the name of the default view via .defaultView property

0.5.3 / 2012-04-16
==================

  * Fixed bug with minimum node version for previous changes that require node 0.6.1 or higher
  * Added limitation to express version installed < 3.0 as this is in alpha and breaks timbits 

0.5.2 / 2012-04-12
==================

  * Fixed bug with making directories in bin/timbits

0.5.1 / 2012-04-12
==================

  * Changes to bin/timbits for CoffeScript v.1.3.x support

0.5.0 / 2012-03-21
==================

  * Unstable experimental release (uses unstable pantry v0.3.x)
  * Exposes pantry for additional configuration and use of optional storage engines

0.4.2 / 2012-03-21
==================

  * Updated package to prevent use of pantry > 0.3.0

0.4.1 / 2012-02-21
==================

  * Added support for new config.base parameter to allow nested timbit servers
  * timbits command line updated to use OS agnostic copy
  * npm init and install removed from command line (doesn't work on windows)


lib/timbits.js

// load required modules
var fs = require('fs')
  , path = require('path')
  , querystring = require('querystring')
  , url = require('url')
  , winston = require('winston')
  , express = require('express')
  , hogan = require('hogan.js')
  , request = require('request');

// default configuration
var config = {
  appName: 'Timbits',
  base: '',
  port: 5678,
  home: process.cwd(),
  maxAge: 60,       // default widget output cache time
  engine: 'hjs',    // default view engine
  discovery: true,  // support automatic discovery via /timbits/json
  help: true,       // allow automatic help pages at /timbits/help and /[name]/help
  test: true,       // allow automatic test pages at /timbits/test and /[name]/test
  json: true,       // allow built in json view at /[name]/json
  jsonp: true       // allow jsonp calls via /[name]/json?callback=
};

// retrieve list of matching files in a folder
function filteredFiles(folder, pattern) {
  var files = [];
    
  if (fs.existsSync(folder)){
    fs.readdirSync(folder).forEach(function(file) {
      if (file.match(pattern) != null)
        files.push(file);
    });
  }
  return files;
}

// automagically load timbits found in the ./timbits folder
function loadTimbits(callback) {

  var folder = path.join(config.home, "/timbits");
  var files = filteredFiles(folder, /\.(coffee|js)$/);
  var pending = files.length;
  
  files.forEach(function(file) {
    var name = file.substring(0, file.lastIndexOf("."));
    timbits.add(name, require(path.join(folder, file)), function() {
      pending--;
      if (pending === 0) callback();
    });
  });
}

// automagically load views for a given timbit
function loadViews(timbit) {
  timbit.views = [];
  
  var pattern = new RegExp('\.' + timbits.app.settings['view engine'] + '$')
    , folder = path.join(config.home, 'views', timbit.viewBase);
    
  if (fs.existsSync(folder)) {
   var files = fs.readdirSync(folder);
   files.forEach(function(file) {
     timbit.views.push(file.replace(pattern, ''));
   });
  }
  
  // We will attempt the default view anyway and hope the timbit knows what it is doing.
  if (timbit.views.length === 0) timbit.views.push(timbit.defaultView);
}

// return a list of possible test values
function getTestValues(values, alltests) {
  if (values != null && values.length != null && values.length !== 0) {
    if (alltests)
      return values;
    else
      return values.slice(0, 1);
  } else {
    return [];
  }
}

// compile built in templates
function compileTemplate(name) {
  var filename = path.join(__dirname, "templates", name + '.hjs');
  var contents = fs.readFileSync(filename);
  return hogan.compile(contents.toString());
}

// generates the allowed methods
function allowedMethods(methods) {
  // default values
  var methodsAllowed = { 'GET': true, 'POST': false, 'PUT': false, 'HEAD': false, 'DELETE': false  };
  // check and override if one of the default methods
  for (var key in methods) {
    var newKey = key.toUpperCase();
    if ( methodsAllowed[newKey]!=undefined) {
      methodsAllowed[newKey] = Boolean(methods[key]);
    }
  }
  return methodsAllowed;
}

var timbits = this;
this.box = {};
this.pantry = require('pantry');
this.templates = {
  help: compileTemplate('help'),
  timbitHelp: compileTemplate('timbit-help'),
  test: compileTemplate('test'),
};

this.getLogLevel = function() {
  switch (process.env.NODE_ENV) {
    case 'production':
      return 'info';
    case 'test':
      return 'error';
    default:
      return 'silly';
  }
};

//added winston object for logging.
this.log = new (winston.Logger)({
  transports: [
    new (winston.transports.Console)({
      colorize: true,
      timestamp: true,
      level: this.getLogLevel()
    })
  ]
});

// creates, configures, and returns a standard express app
this.serve = function(options) {
    
  /* configure options */
  for (var key in options) {
    value = options[key];
    config[key] = value;
  }
  
  /* configure express app */
  var app = timbits.app = express();
   
  app.set('views', "" + config.home + "/views");
  app.set('view engine', config.engine);
  app.set('jsonp callback', config.jsonp);
  
  app.use(express.favicon());
  
  // disable request logging for tests (avoid console clutter)
  if (app.get('env') === 'development')   
    app.use(express.logger('dev'));
    
  app.use(express.compress());
  app.use(express.bodyParser());
  app.use(express.cookieParser());
  app.use(express.static(path.join(config.home, "public")));
  app.use(express.static(path.join(__dirname, "../resources")));
  app.use(express.errorHandler());
  
  // redirect root to help page
  if (config.help) {
    app.all(config.base + "/", function(req, res) {
      res.redirect(config.base + "/timbits/help");
    });
  }
  
  // route json discovery
  app.get(config.base + "/timbits/json", function(req, res) {
    if (config.discovery) {
      res.json(timbits.box);
    } else {
      res.send(404, "Automatic Discovery has been disabled");
    }
  });
  
  // route help page
  app.get(config.base + "/timbits/help", function(req, res) {
    if (config.help) {
      var context = {title: 'Timbits Help', timbits: []};
      for(var key in timbits.box) {
        context.timbits.push(key);
      }
      res.send(timbits.templates.help.render(context));
    } else {
      res.send(404, "Automatic Help has been disabled");
    }
  });
  
  // route master test page
  app.get(config.base + '/timbits/test/:which?', function(req, res) {
    if (config.test) {
      var alltests = (req.params.which === 'all')
        , all_results = []
        , pending = Object.keys(timbits.box).length;
      
      if (pending) {
        for (name in timbits.box) {
          var timbit = timbits.box[name];
          timbit.test('http://' + req.headers.host, alltests, function(results) {
            results.forEach(function(result) {
              all_results.push(result);
            });
            if (--pending === 0) {
              var passed = 0, failed = 0;
              all_results.forEach(function(result) {
                if (result.passed) passed++; else failed++;
              });
              res.send(timbits.templates.test.render({
                title: 'Testing Summary: all timbits',
                passed: passed,
                failed: failed,
                results: all_results         
              }));              
            }
          });
        }
      } else {
        res.send(ck.render(views.test, {}));
      }
    } else {
      res.send(404, "Automatic Test has been disabled");
    }
  });
  
  // automagically load timbits
  loadTimbits(function() {
    try {
       timbits.server = app.listen(process.env.PORT || process.env.C9_PORT || config.port);
       timbits.log.info("Timbits server listening on port " + timbits.server.address().port + " in " + app.settings.env + " mode");
    } catch (err) {
       timbits.log.error("Server could not start on port " + (process.env.PORT || process.env.C9_PORT || config.port) + ". (" + err + ")");
       console.log("\nPress Ctrl+C to Exit");
       process.exit(1);
    }
  });
  
  return app;  
};

// use the 'add' method to place a timbit in the box
this.add = function(name, timbit, callback) {  
  timbits.log.info("Placing " + name + " in the box");
  timbits.box[name] = timbit;
  
  timbit.name = name;
  if (timbit.viewBase == null) timbit.viewBase = name;
  if (timbit.defaultView == null) timbit.defaultView = 'default';
  if (timbit.maxAge == null) timbit.maxAge = config.maxAge;
  timbit.methods = allowedMethods( (timbit.methods == null) ? {} : timbit.methods );
  
  loadViews(timbit);
  
  // route timbit help
  timbits.app.get(config.base + "/" + name + "/help", function(req, res) {
    if (config.help)
      res.send(timbits.templates.timbitHelp.render(timbit));
    else
      res.send(404, "Automatic Help has been disabled");
  });
  
  // route timbit testing
  timbits.app.get(config.base + "/" + name + "/test/:which?", function(req, res) {
    var alltests;
    if (config.test) {
      alltests = req.params.which === 'all';
      timbit.test("http://" + req.headers.host, alltests, function(results) {
        var passed = 0, failed = 0;
        results.forEach(function(result) {
          if (result.passed) passed++; else failed++;
        });
        res.send(timbits.templates.test.render({
          title: 'Testing Summary: ' + timbit.name,
          passed: passed,
          failed: failed,
          results: results         
        }));
      });
    } else {
      res.send(404, "Automatic Test has been disabled");
    }
  });
  
  // main timbit route
  timbits.app.all(config.base + '/' + name + '/:view?', function(req, res) {
    
    // test if the method used is in the allowed methods, 
    if ( timbit.methods[req.method] == undefined || timbit.methods[req.method] == false ) {
      res.send(405, "Method Not Allowed");
	  return;
    } 
    
    // set view name to default view if not specified
    if (req.params.view == null) req.params.view = timbit.defaultView;
    
    // initialize current request context
    var context = {
      name: timbit.name,
      view: timbit.viewBase + '/' + req.params.view,
      maxAge:  timbit.maxAge
    };
    
    // add query string parameters to context
    for (var key in req.query) {
      var has_alias = false;
      
      if (context[key] == null && req.query[key] != null && req.query[key] !== '') {
        
        // handle aliased parameters
        for (var p in timbit.params) {
          if (timbit.params[p] === key) {
            has_alias = true;
            context[p] = req.query[key];
          }
        }
        
        // no alias found
        if (!has_alias)
          context[key] = req.query[key];
      }
    }
    
    // validate request
    for (var key in timbit.params) {
      var param = timbit.params[key];
        
      // if parameter isn't specified, use default value
      if (context[key] == null) context[key] = param.default;
      
      // test provided value based on type
      var value = context[key];
      if (value != null) {
        
        // default parameter type is String
        if (param.type == null) param.type = 'String';
        
        switch (param.type.toLowerCase()) {
          
          case 'number':
            context[key] = Number(value);
            if (isNaN(context[key]))
              throw value + ' is not a valid Number for ' + key;
            break;
            
          case 'boolean':
            switch (value.toLowerCase()) {
              case 'true':
                context[key] = true;
                break;
              case 'false':
                context[key]= false;
                break;
              default:
                throw value + ' is not a valid value for ' + key + '.  Must be true of false';
            }
            break;
            
          case 'date':
            context[key] = Date.parse(value);
            if (isNaN(context[key]))
              throw value + ' is not a valid date for ' + key;
            break;
            
        }
      }
      
      if (param.required && value == null) {
        throw key + " is a required parameter";
      }
      if (value != null && param.strict && param.values.indexOf(value) === -1) {
        throw value + " is not a valid value for " + key + ".  Must be one of [" + (param.values.join()) + "]";
      }
      if (value instanceof Array && !param.multiple) {
        throw key + " must be a single value";
      }
      
     }
     
     // with context created, it's time to consume this timbit
     timbit.eat(req, res, context);
  });
  
  // update example urls if base vpath specified
  if (config.base != null && timbit.examples != null) {
    timbit.examples.forEach(function(example) {
      example.href = config.base + example.href;
    });
  }
  
  // callback after timbit has been loaded
  callback();
  
};

// prototype for Timbit
var Timbit = this.Timbit = function() {};

Timbit.prototype.render = function(req, res, context) {

  // add caching headers
  res.setHeader("Cache-Control", "max-age=" + context.maxAge);
  res.setHeader("Edge-Control", "!no-store, max-age=" + context.maxAge);
  
  if (/^(\w+|(\w+-)+\w+)\/json$/.test(context.view)) {
    if (config.json)
      res.json(context);
    else
      res.send(404, "JSON view has been disabled");
  } else {
    res.render(context.view, context, function(err, str) {
      if (err) {
        timbits.log.error("Error rendering view " + context.view);
        req.next(err);
      } else {
        if (context.callback != null)
          res.json(str);
        else
          res.send(str);
      }
    });
  }
};

Timbit.prototype.fetch = function(req, res, context, options, callback) {

  // use built in render method by default
  if (callback == null) callback = this.render;
  
  var name = options.name || 'data';
  timbits.pantry.fetch(options, function(error, results) {
    if (error) {
      timbits.log.error("Error fetching resource '" + options.uri);
      req.next(error);
    } else {
      if (context[name] != null) {
        if (Object.prototype.toString.call(context[name][0]) === "[object Array]")
          context[name].push(results);
        else
          context[name] = [context[name], results];
      } else {
        context[name] = results;
      }
      callback(req, res, context);
    }
  });
};

Timbit.prototype.eat = function(req, res, context) {
  this.render(req, res, context);
};


Timbit.prototype.generateTests = function(alltests) {
  
  // create combination of required parameters
  var required = [];
  for (var name in this.params) {
    var param = this.params[name];
    if (param.required) {
      var temp = [];
      getTestValues(param.values, alltests).forEach(function(value) {
        if (required.length === 0)
          temp.push(name + "=" + value);
        else
          required.forEach(function(item) {
            temp.push(item + "&" + name + "=" + value);
          });      
      });
      required = temp;
    }
  }
  
  // create list of possible queries using required and optional parameters
  var queries = [];
  required.forEach(function(item) {
    queries.push(item)
  });
  
  // only include optional parameters if all tests are requested
  if (alltests) {
    for (var name in this.params) {
      var param = this.params[name];
      if (!param.required) {
        getTestValues(param.values, alltests).forEach(function(value) {
          if (required.length === 0)
            queries.push(name + "=" + value);
          else
            required.forEach(function(item) {
              queries.push(item + "&" + name + "=" + value);
            });
        });
      }
    }
  }
  
  //create list of testable paths using available views and quiries
  var hrefs = [];
  var name = this.name
  this.views.forEach(function(view) {
    if (queries.length)
      queries.forEach(function(query) {
        hrefs.push("/" + name + "/" + view + "?" + query);
      });
    else
      hrefs.push("/" + name + "/" + view);
  });
  
  return hrefs;
};

Timbit.prototype.test = function(host, alltests, callback) {
  
  // generate dynamic list of test urls
  var tests = this.generateTests(alltests);
  
  // multiplier for  tests total based on allowed methods, determine which methods are used
  var testsLength = 1;
  var getMethod = this.methods["GET"];
  var postMethod = this.methods["POST"];
  if (getMethod && postMethod)
    testsLength = 2;
  
  // add examples to list of tests
  if (this.examples) {
    this.examples.forEach(function(example) {
      tests.push(example.href);
    });
  }
  
  // run each test
  var results = [];
  var name = this.name;
  
  tests.forEach(function(href) {
    if (getMethod) {
      request(host + href, function(error, response, body) {
        error = error || (response.statusCode === 200 ? '' : body);
        results.push({
          timbit: name,
          href: href,
          error: error,
          status: response.statusCode,
          passed: response.statusCode === 200,
          failed: response.statusCode !== 200
       });
      if (results.length === (tests.length * testsLength))
        return callback(results);
      });
    }
    if (postMethod) {
      request.post(host + href, function(error, response, body) {
        error = error || (response.statusCode === 200 ? '' : body);
        results.push({
          timbit: name,
          href: href,
          error: error,
          status: response.statusCode,
          passed: response.statusCode === 200,
          failed: response.statusCode !== 200
       });
      if (results.length === (tests.length * testsLength))
        return callback(results);
      });
    }
  });
  
};

Timbit.prototype.paramsAsArray = function() {
  var array = [];
  for (var key in this.params) {
    array.push({key: key, value: this.params[key]});
  }
  return array;
}

  